src/lib/stepper/stepper.ts

```diff
@@ -0,0 +1,83 @@
+/**
+ * @license
+ * Copyright Google Inc. All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import {CdkStep, CdkStepper} from '@angular/cdk/stepper';
+import {
+  Component,
+  ContentChild,
+  ContentChildren, Directive,
+  // This import is only used to define a generic type. The current TypeScript version incorrectly
+  // considers such imports as unused (https://github.com/Microsoft/TypeScript/issues/14953)
+  // tslint:disable-next-line:no-unused-variable
+  ElementRef,
+  Inject,
+  Optional,
+  QueryList,
+  SkipSelf,
+  ViewChildren,
+  ViewEncapsulation
+} from '@angular/core';
+import {MdStepLabel} from './step-label';
+import {
+  defaultErrorStateMatcher,
+  ErrorOptions,
+  MD_ERROR_GLOBAL_OPTIONS,
+  ErrorStateMatcher
+} from '../core/error/error-options';
+import {FormControl, FormGroupDirective, NgForm} from '@angular/forms';
+import {MdStepHeader} from './step-header';
+
+/** Workaround for https://github.com/angular/angular/issues/17849 */
+export const _MdStep = CdkStep;
+export const _MdStepper = CdkStepper;
+
+@Component({
+  moduleId: module.id,
+  selector: 'md-step, mat-step',
+  templateUrl: 'step.html',
+  providers: [{provide: MD_ERROR_GLOBAL_OPTIONS, useExisting: MdStep}],
+  encapsulation: ViewEncapsulation.None
+})
+export class MdStep extends _MdStep implements ErrorOptions {
+  /** Content for step label given by <ng-template matStepLabel> or <ng-template mdStepLabel>. */
+  @ContentChild(MdStepLabel) stepLabel: MdStepLabel;
+
+  /** Original ErrorStateMatcher that checks the validity of form control. */
+  private _originalErrorStateMatcher: ErrorStateMatcher;
+
+  constructor(mdStepper: MdStepper,
+              @Optional() @SkipSelf() @Inject(MD_ERROR_GLOBAL_OPTIONS) errorOptions: ErrorOptions) {
+    super(mdStepper);
+    this._originalErrorStateMatcher =
```

Comment:
The ternary with a nested || is a little hard to follow. It might be cleaner if it used a regular `if/else`.